File: programs/2.c
#include "../vm/vm.h"

VM __2() {
    __init(256);

    /* function 0: push constants 7 and 35 and return their sum */
    __func(0);
      __push(__I64, 7);
      __push(__I64, 35);
      __add;
      __ret;
    __end;

    /* function 1: push constants 5 and 3 and return their product */
    __func(1);
      __push(__I64, 5);
      __push(__I64, 3);
      __mul;
      __ret;
    __end;

    /* main: call functions, demonstrate if/else and a do-while loop */
    /* call both functions, add their results and print */
    __call(0);
    __call(1);
    __add;
    __print;

    /* if/else demo: condition is true -> prints 100; else would print 200 */
    __push(__I64, 1); /* condition (non-zero = true) */
    __if;
      __push(__I64, 100);
      __print;
    __else;
      __push(__I64, 200);
      __print;
    __end;

    /* while demo (condition-first): check before entering body */
    __push(__I64, 4);   /* initial counter */
    __store;   /* store counter at tape[tp] */

    /* condition: load counter and test */
    __label(cond1);
    __load;
    __while(cond1);
      /* body: load counter, print it, decrement and store back */
      __load;    /* push counter */
      __print;
      __load;    /* push counter */
      __push(__I64, 1);
      __sub;     /* counter - 1 */
      __store;   /* store back */
    __end;

    __halt;

    __fin;
}

File: opt/tmp/raw/2.pl
l1 :-
  const(t0, 7),
  const(t1, 35),
  add(t2, t0, t1),
  ret.

l2 :-
  const(t3, 5),
  const(t4, 3),
  mul(t5, t3, t4),
  ret.

l0 :-
  call(l1, t6),
  call(l2, t7),
  add(t8, t6, t7),
  print(t8),
  const(t9, 1),
  jz(t9, l3),
  const(t10, 100),
  print(t10),
  jmp(l4).

l3 :-
  const(t11, 200),
  print(t11).

l4 :-
  const(t12, 4),
  store(t12).

l5 :-
  load(t13),
  jz(t13, l6).

l7 :-
  load(t14),
  print(t14),
  load(t15),
  const(t16, 1),
  sub(t17, t15, t16),
  store(t17),
  jmp(l5).

l6 :-
  true.
